### ISSUE TITLE
`read_csv`: data type validation for Date/Datetime only in first row

### ISSUE THREAD

**Issue by JulianCologne** [pola-rs/polars#10826]
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

## Dates
```python
# Valid (OK)
pl.read_csv(
    source=StringIO('col\n2023-01-01\n2023-02-01\n2023-03-01'),
    dtypes={'col': pl.Date},
)
┌────────────┐
│ col        │
│ ---        │
│ date       │
╞════════════╡
│ 2023-01-01 │
│ 2023-02-01 │
│ 2023-03-01 │
└────────────┘

# Invalid: row 1 (OK)
pl.read_csv(
    source=StringIO('col\nX\n2023-02-01\n2023-03-01'),
    dtypes={'col': pl.Date},
)
ComputeError: Could not parse `X` as dtype `date` at column 'col' (column number 1).

# Invalid: rows 2,3 (???)
pl.read_csv(
    source=StringIO('col\n2023-01-01\nX\nX'),
    dtypes={'col': pl.Date},
)
┌────────────┐
│ col        │
│ ---        │
│ date       │
╞════════════╡
│ 2023-01-01 │
│ null       │ <<<<<<<<<<<<<<<<<<<<
│ null       │ <<<<<<<<<<<<<<<<<<<<
└────────────┘
```

## Datetime
```python
# Datetime from Date
pl.read_csv(
    source=StringIO('col\n2023-01-01\n2023-02-01\n2023-03-01'),
    dtypes={'col': pl.Datetime},
)
ComputeError: Could not parse `2023-01-01` as dtype `datetime[μs]` at column 'col' (column number 1).

# (1) datetime; (2) date; (3) invalid
pl.read_csv(
    source=StringIO('col\n2023-01-01T01:02:03\n2023-02-01\n2023-03-01'),
    dtypes={'col': pl.Datetime},
)
┌─────────────────────┐
│ col                 │
│ ---                 │
│ datetime[μs]        │
╞═════════════════════╡
│ 2023-01-01 01:02:03 │
│ 2023-02-01 00:00:00 │ <<<<<<<<<<<<<<<<<<<<
│ null                │ <<<<<<<<<<<<<<<<<<<<
└─────────────────────┘
```

## `read_csv` does not respect strict parsing for Date & Datetime after the first row

- only first row seems to use strict parsing and fails on corrupt data

## Datetime parsing additionally really weird

- first row must have a time specified
- other rows do not need it?!?!
- (unfortunately many differenct inconsistent parsing logics across polars...)

### Expected behavior

- Date/Datetime: all rows should use strict parsing
- Datetime: all rows should be parsed equally (why require time on first row?)
- (make parsing logic consistent across polars)

### Installed versions

<details>

```
--------Version info---------
Polars:              0.19.0
Index type:          UInt32
Platform:            Windows-10-10.0.19045-SP0
Python:              3.11.5 (tags/v3.11.5:cce6ba9, Aug 24 2023, 14:38:34) [MSC v.1936 64 bit (AMD64)]

----Optional dependencies----
adbc_driver_sqlite:  <not installed>
cloudpickle:         <not installed>
connectorx:          <not installed>
deltalake:           <not installed>
fsspec:              <not installed>
matplotlib:          <not installed>
numpy:               1.25.2
pandas:              2.0.3
pyarrow:             <not installed>
pydantic:            <not installed>
sqlalchemy:          <not installed>
xlsx2csv:            <not installed>
xlsxwriter:          <not installed>```

</details>

**Comment by MarcoGorelli** [comment]
thanks for the report, will take a look

> other rows do not need it?!?!
(unfortunately many differenct inconsistent parsing logics across polars...)

would appreciate a slightly kinder tone if possible though 🙏

**Comment by JulianCologne** [comment]
@MarcoGorelli closing as this was fixed in #14113
